### DISCUSSION TITLE
How to prevent Mac Mouse Fix Helper from relaunching when quit from terminal?

### DISCUSSION THREAD

**Discussion by YBPN** [noah-nuebling/mac-mouse-fix#865] [Questions]
As an alternative to per-app disabling (which is not yet implemented), I'd like to use Keyboard Maestro to auto-kill/auto-launch `Mac Mouse Fix Helper` depending on which app is active. However, executing:

`killall Mac\ Mouse\ Fix\ Helper`

doesn't work because the helper app just reopens immediately. How to stop it reopening?
(I don't want to have to enable/disable it from the menu bar icon every time.)

**Comment by noah-nuebling** [comment]
Hmm I don't think this is possible at the moment. 

Mac Mouse Fix Helper is registered with macOS's `launchd` service. Launchd tries to always keep Mac Mouse Fix Helper running in the background as long as Mac Mouse Fix is enabled.

The only way I can think of right now to stop the Helper from running is to disable Mac Mouse Fix, aka tell launchd to deregister the Helper. 

Before macOS 13 Ventura I know that you could use the `launchctl` command line tool to start and stop a launchd service. But under macOS 13 Ventura with the new `SMAppService` API, it only seems to be possible to deregister a launchd service not to reregister it. 

Here's how to deregister Mac Mouse Fix Helper from the command line:

```
launchctl bootout gui/501/com.nuebling.mac-mouse-fix.helper
```

But reregistering doesn't seem to be possible from my testing.

A possible solution I could think if is to implement Apple Script support or support for a custom URL scheme that lets other apps tell the Mac Mouse Fix app to register or deregister the Helper. However this would require opening the Mac Mouse Fix app everytime you want to enable/disable the Helper. Maybe you could open it in the background and then quit immediately so it's not noticable? Might be possible to make this nice, but I'd have to look into it. Alternatively I could implement scripting to tell the Helper to stop intercepting any events from you mouse. 

App-specific settings are the next big feature I want to implement, but that might take a while and until them it might pretty useful to build basic scripting support for disabling or enabling Mac Mouse Fix.

Either way, I appreciate your input. It would definitely be fun and cool to support scripting. Would it work for your usecase if you could simply open a url like `macmousefixhelper:suspend` or `macmousefixhelper:resume` to start/stop the helper from intercepting events from you mouse?

**Reply by YBPN** [reply]
Thanks! I might not need the url schema thing because I think I also found a solution before I read your reply (thanks to stackoverflow + trial and error). It looks like it's okay to just run the helper app even if it's not registered in launchd? 

I had to run this once: `launchctl remove com.nuebling.mac-mouse-fix.helper`, after which it was possible to enable/disable the helper any time.

Then what I do in Keyboard Maestro:
When my blacklisted app activates, run shell command: `killall "Mac Mouse Fix Helper"`
When my blacklisted app deactivates, open application: `/Applications/Mac Mouse Fix.app/Contents/Library/LoginItems/Mac Mouse Fix Helper.app`

Seems to work! Hope this helps others.

**Reply by khurshid-alam** [reply]
After removing how do you re-add it ?

**Reply by khurshid-alam** [reply]
> Either way, I appreciate your input. It would definitely be fun and cool to support scripting. Would it work for your usecase if you could simply open a url like `macmousefixhelper:suspend` or `macmousefixhelper:resume` to start/stop the helper from intercepting events from you mouse?

I would like to have this ability to start stop from terminal or apple script. Thanks.
